### PR TITLE
[Merged by Bors] - feat: `privateModule` linter

### PR DESCRIPTION
**Pull request by thorimur**
This linter lints against nonempty modules that have only private declarations, and suggests adding `@[expose] public section` to the top.

This linter found 1 violation in Mathlib, which was fixed in #31822.

---

- [x] depends on: #31823

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+and suggests adding `@[expose] public section` to the top. -/
+register_option linter.privateModule : Bool := {
+  defValue := true
+  descr := "Enable the `privateModule` linter, which lints against nonempty modules that have only \
+    private declarations."
+}
+
+def privateModule : Linter where
+  run stx := do
+    if stx.isOfKind ``Parser.Command.eoi then
```
Comment (adomani) [inline Mathlib/Tactic/Linter/PrivateModule.lean]:
```suggestion
    unless stx.isOfKind ``Parser.Command.eoi do
      return
```
to avoid one indentation level?

Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: In this particular case I actually like that `if` indentation emphasizes that nothing happens unless we're handling an `eoi` token! :) For boilerplate like checking linter options and such, I do like the `unless ... do return` style for the reason you said. 😁

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+-- TODO: `module` is not enabled in MathlibTest yet, so tests should be written for this once it is.
+
+meta section
+
+open Lean Elab Command Linter
+
+/-- The `privateModule` linter lints against nonempty modules that have only private declarations,
+and suggests adding `@[expose] public section` to the top. -/
+register_option linter.privateModule : Bool := {
+  defValue := true
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]: Medium-term, we would like to add this to Mathlib.Init, right? Then this linter should be disabled by default, enabled in the lakefile, and this `register_option` should be public...

Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: Ah, makes sense. I'll do the latter three things here, right? I could also add it to Mathlib.Init here if you think so, or we can battle-test it first.

**Comment by adomani** [comment]
> > By the way, another standard request would be to add a test of this linter! 😄
> 
> I would love to do this, but the module system is not enabled in MathlibTest, so I can't activate the linter (hence the TODO)! 😅 (I had to test locally in a file in regular Mathlib.)

:facepalm: Sorry for not realising that there was a TODO about this in the PR already!

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 2 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: Yes, the latter three can be done here - adding to Mathlib.Init can also happen once this linter is tested properly.

Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: #31823 builds just fine, so I've made this a dependent PR and added a couple of basic tests. :) Should I go ahead and add it to Mathlib.Init, or should we see if we see any edge cases first?

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+`asyncConstsMap.public` and `asyncConstsMap.private` seem to contain exactly the same declarations.
+The difference between public and private declarations is encoded via the presence of the name
+prefix `_private` in the declaration's name. The code here should be robust to potential future
+changes that put public declarations only in `.public` and private ones only in `.private`.
+-/
+
+-- TODO: `module` is not enabled in MathlibTest yet, so tests should be written for this once it is.
+
+public meta section
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]:
```suggestion
meta section
```

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+open Lean Elab Command Linter
+
+namespace Mathlib.Linter
+
+/-- The `privateModule` linter lints against nonempty modules that have only private declarations,
+and suggests adding `@[expose] public section` to the top. -/
+register_option linter.privateModule : Bool := {
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]:
```suggestion
public register_option linter.privateModule : Bool := {
```

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+All new declarations, whether added synchronously with `addDecl` or otherwise,  get added to
+`asyncConstsMap`.
+
+`asyncConstsMap.public` and `asyncConstsMap.private` seem to contain exactly the same declarations.
+The difference between public and private declarations is encoded via the presence of the name
+prefix `_private` in the declaration's name. The code here should be robust to potential future
+changes that put public declarations only in `.public` and private ones only in `.private`.
+-/
+
+-- TODO: `module` is not enabled in MathlibTest yet, so tests should be written for this once it is.
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]: This is obsolete, right?
```suggestion
```

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+/-!
+# Private module linter
+
+This linter lints against nonempty modules that have only private declarations, and suggests adding
+`@[expose] public section` to the top.
+
+## Implementation notes
+
+All new declarations, whether added synchronously with `addDecl` or otherwise,  get added to
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]:
```suggestion
All new declarations, whether added synchronously with `addDecl` or otherwise, get added to
```

**Review by grunweg — COMMENTED** [review]
Thanks! The tests look good to me (I haven't looked at the linter's implementation so far).

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 4 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: I presume we don't care about the linter itself being private. (That's at least what my in-progress PR does.)

Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: Makes sense—I think. I'm still getting my bearings on what should and shouldn't be public. 😅

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 2 above)
Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: [0729c34](https://github.com/leanprover-community/mathlib4/pull/31820/commits/0729c34647f2dc2b68aa9438bf02dbf70a9d12fd) adds the linter to Mathlib.Init, but I can cherry pick this to another branch if I've misunderstood. :)

File: MathlibTest/PrivateModuleLinter/hasOnlyPrivate.lean [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+open Lean
+
+set_option linter.mathlibStandardSet true
+
+theorem foo : True := trivial
+
+def bar : Bool := true
+
+-- Run the linter on artificial `eoi` syntax so that we can actually
```
Comment (Vierkantor) [inline MathlibTest/PrivateModuleLinter/hasOnlyPrivate.lean]: Did you mean something like?
```suggestion
-- Run the linter on artificial `eoi` syntax so that we can capture the output
```

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+def privateModule : Linter where
+  run stx := do
+    if stx.isOfKind ``Parser.Command.eoi then
+      unless getLinterValue linter.privateModule (← getLinterOptions) do
+        return
+      if (← getEnv).header.isModule then
+        let _ := (← getEnv).checked.get -- Wait for everything
+        -- Lint if every declaration is private:
+        if (← getEnv).asyncConstsMap.public.revList.all
+            ((`_private).isPrefixOf ·.constInfo.name) then
```
Comment (Vierkantor) [inline Mathlib/Tactic/Linter/PrivateModule.lean]: There's a function for this in core!
```suggestion
        if (← getEnv).asyncConstsMap.public.revList.all
            (isPrivateName ·.constInfo.name) then
```

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+-/
+module
+
+public meta import Lean.Elab.Command
+public import Lean.Linter.Basic
+public import Lean.Environment
+import all Lean.Environment
+-- Import this linter explicitly to ensure that
+-- this file has a valid copyright header and module docstring.
+import Mathlib.Tactic.Linter.Header
```
Comment (Vierkantor) [inline Mathlib/Tactic/Linter/PrivateModule.lean:13]: I was wondering if this is the right way to import a linter. And if I remove the copyright header here, I don't actually get any error... That's probably an issue for another PR, unless the import is to blame and you know how to fix it.

**Review by Vierkantor — APPROVED** [review]
This PR looks good to me otherwise, the implementation is very nicely explained!

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 9 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean:13]: Good catch! This issue is real, but pre-existing.
- In principle, importing the header linter should suffice for enabling it. However, the linter is disabled by default (to not force it on downstream projects). Mathlib's lakefile enables the mathlibStandardLinter linter set, which is only defined in `Mathlib.Init` (downstream of this file). My guess (untested) is that this means the linter is not activated here yet. Let me try this myself.
- In practice, the doc-string linter does not fire at all. That's a bug to be fixed. It also doesn't fire if I make this not a module. Let me check if this is a recent regression or not...
- I would hope that the way we import the linter doesn't matter here...

Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean:13]: I tested in a file importing Mathlib.Init: everything still works.

Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean:13]: I tested this on v4.25.0, with the same result: mutilating the copyright header in `Lean/Json.lean` (importing Mathlib.Init) is detected fine, tinkering with the one in `DeprecatedSyntaxLinter.lean` is not.

My current theory is that the switch to linter sets broke this; let me test that...

Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean:13]: Bingo: editing the `lakefile.lean` to modify `mathlibOnlyLinters` as follows
```
abbrev mathlibOnlyLinters : Array LeanOption := #[
  ⟨`linter.mathlibStandardSet, true⟩,
  ⟨`linter.style.header, true⟩,
  -- remaining options
```
means edits to the copyright header are still picked up.

A band-aid fix would be to add `linter.style.header` to the lakefile, with a comment why it is required.
I'm not sure if there should be a deeper fix related to linter sets. What do you think?

File: MathlibTest/PrivateModuleLinter/hasOnlyPrivate.lean (hunk 7 above)
Comment (thorimur) [reply MathlibTest/PrivateModuleLinter/hasOnlyPrivate.lean]: 🤦 Thanks for catching this, somehow I deleted the words "guard the message"...!

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 8 above)
Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: Ah, nice catch! :) Also, I received a reply from Sebastian on [the Zulip thread](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Mathlib.20has.20moved.20to.20the.20new.20module.20system/near/558371376); apparently we ought to still just filter by private names in `env.constants.map₂` here, no change necessary. 🙃

But it is a persistent hash map, and I'm wondering if that's as performant as this approach, where `.all` short-circuits? Maybe modules contain too few local declarations for this to matter, but I'm going to bench the current approach just in case, and compare. :)

(Note: the PersistentHashMap `forIn` uses an `ExceptT` `foldlM`, which is still a fold, so I'm suspicious of that...)

**Comment by thorimur** [comment]
!bench

**Comment by Vierkantor** [comment]
The variations seem so small, I don't think we need to worry about it. Also, thinking about it theoretically: we're not going to have a million declarations in a module, so there *should* not be a problem if we stop short-circuiting and doing a few extra string comparisons.

If you want to be really sure, though, the only way to re-benchmark is to make an (empty) commit and run `!bench` again. (Unsure how it works for this new Radar though.)

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      if (← getEnv).header.isModule then
+        -- Don't lint an imports-only module:
+        if !(← getEnv).constants.map₂.isEmpty then
+          -- Exit if any declaration is public:
+          for (decl, _) in (← getEnv).constants.map₂ do
+            if !isPrivateName decl then return
+          -- Lint if all names are private:
+          let topOfFileRef := Syntax.atom (.synthetic ⟨0⟩ ⟨0⟩) ""
+          logLint linter.privateModule topOfFileRef
+            "The current module only contains private declarations.\n\n\
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]: Perhaps one newline suffices?
```suggestion
            "The current module only contains private declarations.\n\
```

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+/-- The `privateModule` linter lints against nonempty modules that have only private declarations,
+and suggests adding `@[expose] public section` to the top. -/
+public register_option linter.privateModule : Bool := {
+  defValue := false
+  descr := "Enable the `privateModule` linter, which lints against nonempty modules that have only \
+    private declarations."
+}
+
+def privateModule : Linter where
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]: Does the docBlame linter not complain about this? Can you add a doc-string, please? (This can mostly copy-paste the option command, or even use `inherit_doc`.)

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  run stx := do
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]: For the other linters, we used
```suggestion
def privateModule : Linter where run stx := do
```
(and then you can unindent the lines below)

File: Mathlib/Tactic/Linter/PrivateModule.lean [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        -- Don't lint an imports-only module:
+        if !(← getEnv).constants.map₂.isEmpty then
+          -- Exit if any declaration is public:
+          for (decl, _) in (← getEnv).constants.map₂ do
+            if !isPrivateName decl then return
+          -- Lint if all names are private:
+          let topOfFileRef := Syntax.atom (.synthetic ⟨0⟩ ⟨0⟩) ""
+          logLint linter.privateModule topOfFileRef
+            "The current module only contains private declarations.\n\n\
+            Consider adding `@[expose] public section` at the beginning of the module."
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/PrivateModule.lean]: Not sure if we want to suggest this already, but we could also
```suggestion
            Consider adding `@[expose] public section` at the beginning of the module,\
            or selectively marking declarations as public."
```

**Review by grunweg — COMMENTED** [review]
Thanks! Just two style nits (and a question about wording).

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 10 above)
Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: I initially tried zero newlines and one newline, but two looked the least busy to me in the infoview. I don't mind changing it if you feel strongly, but I do deliberately like it as is. :)

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 11 above)
Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: 🤦 Oops, simply forgot this. Added some info here about linter's behavior re: syntax and logging.

> Does the docBlame linter not complain about this?

No, it doesn't complain (perhaps because it's technically `private` now? Or simply not imported? 🤔).

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 13 above)
Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: Ooh, nice! This message feels a lot better. I've committed it. :)

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 10 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: I'd rate my preference as "medium". Having a newline is unusual compared to the other linter errors --- but this might be subjective, so I'll let you decide :-)

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 11 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: I guess the declaration being private is the reason. (Would you like to check if it still errors on public declarations? Otherwise, that would be a bad bug to silently have, even if unrelated to this PR.)

File: Mathlib/Tactic/Linter/PrivateModule.lean (hunk 10 above)
Comment (thorimur) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: Ok—having looked at it again before I responded here, I do still like two newlines, if that's alright. 😁 I might even go so far as to suggest that other linters should not shy away from newlines when it breaks up a block of text into different roles! :) (With roles here being: what's wrong; what you should do; and then, of course, how to turn this linter off, added by `logLint`, which already separates itself using two newlines.)

But yes, ultimately subjective, I imagine. I can imagine someone feeling that multiple blocks is *more* busy, even though I feel differently.

Comment (grunweg) [reply Mathlib/Tactic/Linter/PrivateModule.lean]: Fair: your opinion is well-justified also, so I'm happy to go with what we currently have.
(I agree about the separation of roles in principle --- but given that each "role" is about one line, having the newlines feels more hindering than helpful. Again, we can disagree on the particular preferences.)

**Review by grunweg — COMMENTED** [review]
Thanks! Let's iterate on any refinements in future PRs.
bors merge

**Comment by grunweg** [comment]
Let's try again:
bors merge

**Comment by grunweg** [comment]
I don't see any obvious clues, so let me ask on zulip.
